### PR TITLE
Add support for modules

### PR DESCRIPTION
**Pull request by nechaido**
BREAKING:

- Removed support for `Projection` decorator;
- Duplicate definition of a domain will now throw an error instead of
  reassigning;
- `metaschemas.fs.load` now accept `isRoot` flag;
- `ms.views`, `ms.forms` and `ms.actions` are removed from metaschema instance,
  they are now stored on a relevant category:
  ```js
  ms.categories.get("name") ===
    {
      name,
      definition,
      views,
      forms,
      actions,
      references,
    };
  ```
- In views renamed `Category` -> `Base`, it is optional if it is not provided,
  the name of the category the view is defined upon will be used instead;
- `view.definition.definition` is now stored in `view.Fields` reassigning
  original array;
- Every schema type now has a corresponding file extension: `action`, `domains`,
  `category`, `form`, `view`;
- Every action is now removed from category definition.

MINOR:

- Automatically assign empty `Fields` object to forms that have none;
- Schemas can now be stored in subfolders, although every schema would still be
  loaded globally as if it was located in a root folder;
- Actions can now be defined in a separate file, in that case, the category will
  be computed this way:

  - If filename has following structure `categoryName.actionName.action`,
    `categoryName` will be used;
  - if the filename is `actionName.action` name of a module, the file is located
    in.

  The same naming convention is used for views and forms;

### CODE REVIEW

File: lib/schema-fs-loader.js [hunk 1]
```diff
@@ -115,8 +123,96 @@ const loadAndCreate = (dir, context, callback) => {
   }
 };
 
+const getSortOrder = file => {
+  if (file.includes('.')) {
+    return 0;
+  }
+  if (file.endsWith('.category')) {
```
Comment (SemenchenkoVitaliy) [inline lib/schema-fs-loader.js]: This statement will never execute because file will have `.` and function will exit with the first statement.

**Comment by tshemsedinov** [comment]
- Category name is globally unique, so if we have folder `Person` and file `Person.category` in different places we need to show error message as well if we have multiple `Person` folders or multiple `Person.category` files in directories tree.
- Folder can be cased as category, but schemas root folder can't be casted as category.

File: lib/schema-fs-loader.js [hunk 2]
```diff
@@ -37,43 +39,89 @@ const loadSchema = (filepath, context, callback) => {
...
+        schemas.actions.push(obj);
+      }
+    }
+
+    schemas.sources.push({ path: filepath, source });
+    callback(null);
   });
 };
 
-// Load schemas from directory
```
Comment (belochub) [inline lib/schema-fs-loader.js]: Why did you delete these comments?

File: lib/schema-fs-loader.js [hunk 3]
```diff
@@ -37,43 +39,89 @@ const loadSchema = (filepath, context, callback) => {
...
+const load = (dir, context, schemas, callback) => {
+  fs.readdir(dir, (error, files) => {
+    if (error) {
+      callback([error]);
       return;
     }
-    files = files.map(file => path.join(dir, file));
-    loadFiles(files, context, callback);
+
+    const sortedFiles = files.sort(sorter);
```
Comment (belochub) [inline lib/schema-fs-loader.js]: You don't need to create a new variable, `Array.prototype.sort()` sorts in place.
```suggestion
    files.sort(sorter);
```

File: lib/schema-fs-loader.js [hunk 4]
```diff
@@ -37,43 +39,89 @@ const loadSchema = (filepath, context, callback) => {
...
-    return;
-  }
-  if (!dir.includes(path.sep)) dir = path.join(schemasDir, dir);
-  fs.readdir(dir, (err, files) => {
-    if (err) {
-      callback(err);
+const load = (dir, context, schemas, callback) => {
+  fs.readdir(dir, (error, files) => {
+    if (error) {
+      callback([error]);
```
Comment (belochub) [inline lib/schema-fs-loader.js]: Why is the `error` being wrapped in the array here?

File: lib/decorators.js [hunk 5]
```diff
@@ -122,7 +122,6 @@ class History extends Decorator {}
 // Not materialized entity decorators
 
 class View extends Decorator {}
-class Projection extends Decorator {}
```
Comment (belochub) [inline lib/decorators.js]: Is the `Projection` decorator removal relevant to the modules support?

File: lib/schema-fs-loader.js [hunk 6]
```diff
@@ -84,19 +132,40 @@ const load = (dir, context, callback) => {
...
+      if (err) {
+        callback(err);
+      } else {
+        callback(...create(schemas));
       }
+    });
+  };
+
+  if (Array.isArray(dir)) {
+    metasync.map(
```
Comment (belochub) [inline lib/schema-fs-loader.js]: Why use `metasync.map()` here, if you are not using the resulting array in the callback afterward?
```suggestion
    metasync.each(
```

File: lib/schema.js [hunk 7]
```diff
@@ -455,47 +175,10 @@ class Metaschema {
   constructor() {
     this.domains = new Map();
     this.categories = new Map();
-    this.forms = new Map();
-    this.views = new Map();
-    this.actions = new Map();
+    //this.views = new Map();
```
Comment (belochub) [inline lib/schema.js]: This is supposed to be removed, I guess.

File: test/schemas/view/Person.View.view [hunk 8]
```diff
@@ -9,3 +9,4 @@ View({
     { Age: Person => Date.now() - Person.DOB },
   ],
 })
+
```
Comment (belochub) [inline test/schemas/view/Person.View.view]: Redundant empty line.

File: lib/schema.js [hunk 9]
```diff
@@ -672,14 +355,14 @@ class Metaschema {
   //   args - <Object>, action arguments
   // Returns: <MetaschemaError> | <null>
   validateAction(category, actionName, args) {
-    const categoryActions = this.actions.get(category);
+    const categoryActions = this.categories.get(category);
```
Comment (belochub) [inline lib/schema.js]: It's not `categoryActions`, it's `category` itself here. We can either change the variable name or extract the actions here:
```suggestion
    const categoryActions = this.categories.get(category).actions;
```

File: lib/schema.js [hunk 10]
```diff
@@ -713,86 +396,94 @@ class Metaschema {
...
 
 // Creates Metaschema instance
 //   schemas - <Iterable> schemas in form [name, schema, source]
 //             (the 'source' is optional)
 // Returns: [<MetaschemaError>, <Metaschema>]
 const create = schemas => {
-  const ms = new Metaschema();
-  for (const [name, schema, source] of schemas) {
-    ms[addSchema](name, schema, source);
+  const ms = new Metaschema(schemas);
```
Comment (belochub) [inline lib/schema.js]: Should we maybe split this function into the multiple smaller ones?

**Comment by belochub** [comment]
At the moment, there is also duplicate information in `View`s, the category name is specified both in the filename and in the `Category` field, e.g.:
https://github.com/metarhia/metaschema/blob/44a0ce5205326e29c57927eefb27efec2d1ff79f/test/schemas/view/Person.View.view#L2

File: lib/schema-processors.js [hunk 11]
```diff
@@ -0,0 +1,268 @@
...
+        });
+      }
+    }
+  }
+  return errors.length ? new MetaschemaError(errors) : null;
+};
+
+const processFields = (ms, category, fields, info) => {
+  const errors = [];
+  for (const key in fields) {
```
Comment (lundibundi) [inline lib/schema-processors.js]: `Object.entries`?

File: lib/schema-processors.js [hunk 12]
```diff
@@ -0,0 +1,268 @@
...
+      property: 'Returns',
+    })
+  );
+
+  const formName = def.Form || action.name;
+  const form = category.forms.get(formName);
+
+  if (form) {
+    Object.keys(def.Args).forEach(arg => {
+      if (form.definition.Fields[arg]) {
```
Comment (lundibundi) [inline lib/schema-processors.js]: I'd prefer `common.iter` + `filter` here.

File: lib/decorators.js (hunk 5 above)
Comment (nechaido) [reply lib/decorators.js]: We would need to add support for `.projection` file extension, and remove it in a followup PR.

File: lib/schema.js [hunk 13]
```diff
@@ -398,20 +397,18 @@ class Metaschema {
...
-
+const addDomains = (ms, schemas) => {
   const errors = [];
 
-  for (const { definition: domains } of schemas.domains) {
+  for (const { definition: domains } of schemas) {
     for (const [name, domain] of Object.entries(domains)) {
       if (ms.domains.has(name)) {
-        // TODO: Handle Error
-        continue;
```
Comment (nechaido) [inline lib/schema.js]: Restore `continue`.

File: lib/schema-fs-loader.js [hunk 14]
```diff
@@ -37,43 +40,112 @@ const loadSchema = (filepath, context, callback) => {
...
+    schemas.sources.push({ path: filepath, source });
+    callback(null);
   });
 };
 
 // Load schemas from directory
 //   dir - <string> | <null> directory to load schemas from
 //   context - <Object> object to be assigned to global during loading
+//   schemas - <Object> object to collect schemas to
+//         categories - <Object[]>
```
Comment (belochub) [inline lib/schema-fs-loader.js]: Incorrect indentation here and below.
```suggestion
//     categories - <Object[]>
```

File: lib/schema-processors.js [hunk 15]
```diff
@@ -0,0 +1,268 @@
...
+const { SchemaValidationError, MetaschemaError } = require('./schema-errors');
+
+// Verifies that there could be link from source category to destination
+//   source - <Object>
+//   sourceName - <string>
+//   destination - <Object>
+//   destinationName - <string>
+//   propertyName - <string>
+// Returns: <SchemaValidationError> | <null> information about error or null
+//          if link is valid
```
Comment (belochub) [inline lib/schema-processors.js]: Incorrect indentation:
```suggestion
//     if link is valid
```

File: lib/schema-errors.js [hunk 16]
```diff
@@ -13,7 +13,8 @@ const schemaValidationErrorSerializers = {
     ` '${error.info.property}' from ${error.source}.${error.property}`,
   duplicateName: error => `Attempt to define field '${error.property}' ` +
     `in ${error.source} which is already defined in ${error.info.location}`,
-  unlinkedForm: error => `Form '${error.source}' is not linked to a category`,
+  duplicate: error => `Duplicate ${error.info.entity} '${error.source}'`,
+  unlinked: error => `Unlinked ${error.info.entuty} '${error.source}'`,
```
Comment (belochub) [inline lib/schema-errors.js]: Typo:
```suggestion
  unlinked: error => `Unlinked ${error.info.entity} '${error.source}'`,
```

File: lib/schema-fs-loader.js [hunk 17]
```diff
@@ -25,84 +27,185 @@ const loadSchema = (filepath, context, callback) => {
...
+    const name = path.basename(filepath, ext);
     let exports;
     try {
       exports = processSchema(name, source, context);
     } catch (e) {
       callback(e);
       return;
     }
-    callback(null, name, exports, source);
+    callback(null, name, ext.slice(1), exports, source);
```
Comment (belochub) [inline lib/schema-fs-loader.js]:
```suggestion
    callback(null, name, common.fileExt(ext), exports, source);
```

File: lib/schema-fs-loader.js [hunk 18]
```diff
@@ -25,84 +27,185 @@ const loadSchema = (filepath, context, callback) => {
...
+
+          if (stats.isDirectory()) {
+            load(filePath, context, false, cb);
+          } else {
+            const moduleName = isRoot ? null : path.basename(dir);
+            loadModuleSchema(filePath, context, moduleName, cb);
+          }
+        });
+      },
+      error => callback(error, schemas)
```
Comment (belochub) [inline lib/schema-fs-loader.js]:
```suggestion
      error => {
        callback(error, schemas);
      }
```

File: lib/schema-fs-loader.js [hunk 19]
```diff
@@ -25,84 +27,185 @@ const loadSchema = (filepath, context, callback) => {
...
+    categories: [],
+    domains: [],
+    views: [],
+    forms: [],
+    actions: [],
+    sources: [],
+  };
+
+  const process = schemasArray =>
+    schemasArray.forEach(([type, schema]) =>
```
Comment (belochub) [inline lib/schema-fs-loader.js]:
```suggestion
    schemasArray.forEach(([type, schema]) => {
          schemas[typeToPlural[type]].push(schema);
    });
```

File: test/validate.js [hunk 20]
```diff
@@ -12,317 +13,360 @@ const validateTest = metatests.test('validate');
...
-            new ValidationError('invalidType', 'Schema1.DoSome.Name', {
-              expected: 'string',
-              actual: 'number',
-            }),
-          ]),
+metasync.each(
+  [metaschemaPath, schemaPath],
+  (dir, cb) =>
+    metaschema.fs.load(dir, null, true, (err, arr) => {
+      arr.forEach(([type, schema]) => schemas[typeToPlural[type]].push(schema));
```
Comment (belochub) [inline test/validate.js]:
```suggestion
      arr.forEach(([type, schema]) => {
        schemas[typeToPlural[type]].push(schema);
      });  
```

File: lib/schema-fs-loader.js [hunk 21]
```diff
@@ -170,16 +170,22 @@ const loadAndCreate = (dir, context, callback) => {
     sources: [],
   };
 
+  const process = schemasArray =>
```
Comment (lundibundi) [inline lib/schema-fs-loader.js]: Perhaps `fillSchemas` or `fillSchemasWith`?
